### PR TITLE
Ck/10812 document list editing

### LINKED ISSUE #10812: [Document lists] Introducing DocumentListEditing
Part of #2973.

Implement the editing part of the document lists feature:

*   Schema
*   Converters (upcast, downcast)
*   Support for:
    *   indent
    *   type
*   Error when loading DocLists and PlainLists/ToDoList/PlainListStyle together

### LINKED ISSUE #11065: [Document lists] Research challenges in porting list properties feature to document lists
The goal is to unblock the refinement of the final (implementation) part of the ticket.

### PR DESCRIPTION
**Pull request by niegowski**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Feature (list): Introducing the document list editing feature. Closes #10812.

Other (engine): The `isAllowedInsideAttributeElement` option was removed, from now on `AttrubuteElements` are allowed to wrap any view element.

Fix (link): The link decorators should be converted on the block images only once (should not wrap block image with an additional link).

Other (engine): The `Schema` is extended by the new generic type: `$container`. Container elements can contain `$block` or `$container` elements. 

Other (engine): The `DowncastHelpers` are passing an additional parameter to the creator functions (the `data` that provides more context to the element creator callback.

Other (engine): The `ConversionApi` provided by the `UpcastDispatcher` is extended with an additional method `keepEmptyElement` that marks an element that was created during splitting some model element that should not get removed on conversion even if it's empty. 

Other (engine): The `Differ` change entries for `insert` and `remove` types are extended with a map of attributes that were set while inserting an element or that were on an element while it got removed.

MINOR BREAKING CHANGE (engine): The `isAllowedInsideAttributeElement` option is removed so `AttributeElement`s can wrap any view element (according to positions). Make sure that you are not wrapping any `ContainerElement` by an accident by not checking the target in the converter. Those would previously get wrapped by an `AttributeElement` that immediately would be removed by the `ContainerElement` within it so there would not be any visual effect.

Feature (list): Introducing the document list properties feature. Closes #11065.

Internal (engine): Added option for the `DomConverter` to transparently render only the content of the element in the data pipeline.

---

### Additional information

### CODE REVIEW

File: packages/ckeditor5-list/src/documentlist/converters.js [hunk 1]
```diff
@@ -0,0 +1,521 @@
...
+						findAndAddListHeadToMap( entry.range.start.getShiftedBy( 1 ), itemToListHead );
+						refreshItemParagraphIfNeeded( item, [] );
+					} else {
+						changedItems.add( item );
+					}
+				} else if ( item.hasAttribute( 'listItemId' ) ) {
+					refreshItemParagraphIfNeeded( item );
+				}
+			}
+		}
```
Comment (oleq) [inline packages/ckeditor5-list/src/documentlist/converters.js:147]: Let's make a named helper out of it.

File: packages/ckeditor5-list/src/documentlist/converters.js [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+
+					// Don't need to iterate further if we already know that the item is wrapped appropriately.
+					if ( stackIdx < 0 ) {
+						return;
+					}
+				}
+			}
+
+			itemsToRefresh.add( item );
+		}
```
Comment (oleq) [inline packages/ckeditor5-list/src/documentlist/converters.js:247]: Do we need to create all these helpers every time `change:data` kicks in? I'd extract them somewhere else.

**Review by oleq — CHANGES_REQUESTED** [review]
1. Let's merge it to a new feature branch instead of master.
2. Because the indent command branch contains improvements and refactoring of some fragments of code mentioned in this review, let's apply changes there and avoid doing the same thing twice.

File: packages/ckeditor5-html-support/tests/htmlcomment-integration.js [hunk 3]
```diff
@@ -387,9 +387,9 @@ describe( 'HtmlComment integration', () => {
 			expect( editor.getData() ).to.equal(
 				'<!-- c1 -->' +
 				'<figure class="image">' +
-					'<img src="/assets/sample.png" alt="Example image">' +
 					'<!-- c3 -->' +
 					'<!-- c2 -->' +
+					'<img src="/assets/sample.png" alt="Example image">' +
```
Comment (Reinmar) [inline packages/ckeditor5-html-support/tests/htmlcomment-integration.js:392]: This test didn't work on the feature branch, but we didn't notice  this so far. So, it's unrelated to this PR.

File: packages/ckeditor5-link/src/linkediting.js [hunk 4]
```diff
@@ -193,8 +193,12 @@ export default class LinkEditing extends Plugin {
 
 			editor.conversion.for( 'downcast' ).attributeToElement( {
 				model: decorator.id,
-				view: ( manualDecoratorName, { writer } ) => {
-					if ( manualDecoratorName ) {
+				view: ( manualDecoratorValue, { writer, schema }, { item } ) => {
+					if ( !schema.isInline( item ) ) {
```
Comment (Reinmar) [inline packages/ckeditor5-link/src/linkediting.js:197]:
```suggestion
					// Manual decorators for block links are handled e.g. in LinkImageEditing.
					if ( !schema.isInline( item ) ) {
```

File: packages/ckeditor5-link/src/utils/automaticdecorators.js [hunk 5]
```diff
@@ -66,6 +66,11 @@ export default class AutomaticDecorators {
 				if ( !conversionApi.consumable.test( data.item, 'attribute:linkHref' ) ) {
 					return;
 				}
+
+				if ( !( data.item.is( 'selection' ) || conversionApi.schema.isInline( data.item ) ) ) {
```
Comment (Reinmar) [inline packages/ckeditor5-link/src/utils/automaticdecorators.js:70]:
```suggestion
				// Automatic decorators for block links are handled e.g. in LinkImageEditing.
				if ( !( data.item.is( 'selection' ) || conversionApi.schema.isInline( data.item ) ) ) {
```

**Comment by CatStrategist** [comment]
@ckeditor/qa-team 

First milestone for Document Lists is ready for tests. You can access it on `ck/10812-document-list-editing` branch in list's package manual test `document-list`.

Ready and most important things to test are:
- Creating bulleted/numbered lists
- Multi-block list items
- Mixing bulleted/numbered lists
- Handling enter key and creating new list items
- Handling delete/backspace
- Indenting/Outdenting list items through UI Button and Tab/Shift+Tab keys
- List navigation
- Order of commands' execution. E.g. table as a list item and selection is inside table's cell - table tab handler should react to keystroke. (Whatever is closest to selection).
- Setting/getting document lists through setData/getData
- Verify correct behaviour when working with multiple lists, e.g. merge two lists
- Inserting and behaviour of widgets as a list item. Currently you can't insert widgets directly as a list item, it's a known bug. To make widget a list item you must first create it out of a list and then make it a list with bulleted/numbered list button.
- Verify if old lists aren't broken
- Test nesting lists in a block quote/tabels 
- Test nesting block quote/tabels in a list

Except for finding bugs we would welcome feedback on UX. Please share any ideas you might have about improving behaviour of a document lists as well as anything that bothered you or seems off. :)

If you have any questions please let me know!

**Comment by CatStrategist** [comment]
@Mgsy Please link all related bugs in this pull request.

**Comment by Mgsy** [comment]
## Plugins conflict

It's impossible to use `DocumentList` feature along with `TodoList` and `ListProperties` plugins, as these require `ListEditing` resulting in conflict between `DocumentListEditing`.

**Comment by FilipTokarski** [comment]
Cannot remove a widget which is a list item - https://github.com/ckeditor/ckeditor5/issues/11346

**Comment by FilipTokarski** [comment]
List items are not removed correctly with backspace - https://github.com/ckeditor/ckeditor5/issues/11347

**Comment by FilipTokarski** [comment]
Error when removing formatting on text - https://github.com/ckeditor/ckeditor5/issues/11348

**Comment by FilipTokarski** [comment]
Error after getting data with html comment in list item - https://github.com/ckeditor/ckeditor5/issues/11350

**Comment by LukaszGudel** [comment]
Image is inserted incorrectly with drag&drop into paragraph in list - https://github.com/ckeditor/ckeditor5/issues/11351

**Comment by LukaszGudel** [comment]
Removing widget breaks the document list - https://github.com/ckeditor/ckeditor5/issues/11352

**Comment by FilipTokarski** [comment]
Cannot insert text before or after a widget inside a list item - https://github.com/ckeditor/ckeditor5/issues/11353

**Comment by LukaszGudel** [comment]
Error after getting data with code block in list - https://github.com/ckeditor/ckeditor5/issues/11354

**Comment by FilipTokarski** [comment]
List gets split when widget is pasted/dropped into the list item - https://github.com/ckeditor/ckeditor5/issues/11355

**Comment by FilipTokarski** [comment]
List items misaligned and covered after aligning table to the left - https://github.com/ckeditor/ckeditor5/issues/11361

**Comment by LukaszGudel** [comment]
I've created a single ticket for our UX feedback. We will add new comments there - https://github.com/ckeditor/ckeditor5/issues/11371

**Comment by dufipl** [comment]
Image with captions split list: https://github.com/ckeditor/ckeditor5/issues/11374

**Comment by dufipl** [comment]
Blockquote does not work with document-list https://github.com/ckeditor/ckeditor5/issues/11383

**Comment by dufipl** [comment]
Drag and dropping or pasting the table into the table cell with document list inside clears cell content: https://github.com/ckeditor/ckeditor5/issues/11384

**Comment by dufipl** [comment]
Media embed does not work with document-list: https://github.com/ckeditor/ckeditor5/issues/11386

**Comment by niegowski** [comment]
Closing this PR as code moved to epic branch #11503

**Pull request closed without merging**